Update the rcljava parameter APIs

clalancette opened the pull request:
The main goal of this PR is to bring feature parity to the parameter APIs inside of rcljava.  The parameter APIs in rclcpp and rclpy were changed fairly significantly during the Dashing release cycle, and have been furthered refined in Foxy.  This PR brings rcljava up to the same level of API compliance.  There are also tests to confirm the behavior.

I know this is a very large PR, and hence may be difficult to review.  I have made an attempt to split the changes up into logical commits, so my suggestion is to review commit-by-commit, rather than looking at the whole thing at once.

@jacobperron FYI

clalancette committed:
Implement all parameter types.

This implementation was missing BoolArray, IntegerArray,
DoubleArray, and StringArray.  Doing this required a change
to the constructor for ParameterVariant, since List is a
Generic and we can't have multiple constructors with the
same signature.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Update the Node parameter methods to Dashing/Foxy equivalents.

The parameter API was significantly updated during the Dashing
release cycle.  Update the API in rcljava to provide similar
functionality and make the API look a lot more like rclcpp
and rclpy.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Add log4j initialization to all tests.

It gets rid of a warning when starting the tests that looks like:

log4j:WARN No appenders could be found for logger (org.ros2.rcljava.common.JNIUtils).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Add tests for Node parameters.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Add node parameters test with undeclared parameters.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Add in another ParameterVariant.

This one just takes a name and sets the type to NOT_SET.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Make the parameter separator a class variable.

It doesn't need to change, so it can just be a class variable.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Store the parameter and the descriptor.

That way we have the information available when other methods
want to retrieve it.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Switch parameters to using map lookups.

The parameters are stored in a HashMap of name -> parameter,
but we were unnecessarily iterating over the HashMap to find
things.  Instead, look up the items directly in the map which
should be much faster.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Plumb undeclaredParameters through the Node constructor.

Whether to allow undeclared parameters is a decision that is
made during Node creation.  Plumb through the necessary option
so that the user can choose to allow undeclared parameters when
they create the node.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

clalancette committed:
Add in a ParameterCallback interface.

This interface is what users will have to implement in order
to have their callback called when parameters are set.

Signed-off-by: Chris Lalancette <clalancette@openrobotics.org>

jacobperron commented on rcljava/src/main/java/org/ros2/rcljava/node/Node.java:
```diff
@@ -175,15 +177,39 @@
 
   String getName();
 
+  void declareParameter(ParameterVariant parameter);
```
It would be nice to add some docblocks to describe these newly added functions.

clalancette replied:
Yeah, good call.  Done in 36daa72840b68189f0c6d7d763440c2395ec0d46

jacobperron commented on rcljava/src/main/java/org/ros2/rcljava/node/Node.java:
```diff
@@ -175,15 +177,39 @@
 
+  void declareParameter(ParameterVariant parameter);
+
+  void declareParameter(ParameterVariant parameter, rcl_interfaces.msg.ParameterDescriptor descriptor);
```
Have you considered creating a new parameters interface instead of adding these methods to the Node interface?

We can then have the Node interface extend the parameters interface, which I think is a little nicer for extensibility.

Taking it a step further, we could move the implementation to a `ParametersImpl` class that `NodeImpl` uses.  I think this is also similar to what we have in rclcpp.

clalancette replied:
I followed the example of both rclcpp and rclpy in putting these methods directly into the main Node class.  I think consistency is nice to have here, as it means looking across all of the rcl* layers is easier.

That being said, I don't feel really strongly about this, so if you still think we should refactor into a ParametersImpl class, I can look into doing that.

jacobperron replied:
I think this *is* what rclcpp is doing (I guess not rclpy); composition of interfaces. We'd still have these methods here, but the implementation would be in another object (e.g. for re-use in a future LifecycleNode or something).

jacobperron replied:
It it seems like too much overhead we postpone a refactor until we see a direct need for it.

jacobperron commented on rcljava/src/main/java/org/ros2/rcljava/node/NodeImpl.java:
```diff
@@ -392,103 +401,299 @@ public final String getName() {
+
+  public void declareParameter(ParameterVariant parameter, rcl_interfaces.msg.ParameterDescriptor descriptor) {
+    synchronized (parametersMutex) {
+      internalDeclareParameter(parameter, descriptor);
```
Small suggestion: If we move the implementation to it's own class, `ParametersImpl`, we can probably simplify a bit and use the object intrinsic lock. E.g. make `internalDeclareParameter` a synchronized method and other places where we don't want to synchronize the whole method, use `synchronize (this)` ([ref](https://docs.oracle.com/javase/tutorial/essential/concurrency/locksync.html)).

I think it would make the code a little more readable as we wouldn't have to worry about making sure we wrap `internalDeclareParameter` in a lock every time we use it. Alternatively, we can move this lock inside `internalDeclareParameter`.

clalancette replied:
So I've actually refactored this in 36daa72 to be a lot nicer.  Basically I changed it so that the singular methods always call the plural methods.  Then I take the lock in the plural methods.  That way, we get the atomicity we want, without having the internal method.  The one exception to this is `setParameter/setParameters/setParametersAtomically`, because there we actually do want the locking to be different depending on which variant we are calling.  Take a look at the new code and let me know what you think.

jacobperron replied:
LGTM

jacobperron commented on rcljava/src/main/java/org/ros2/rcljava/node/NodeImpl.java:
```diff
@@ -392,103 +401,299 @@ public final String getName() {
+    }
+    Iterator<ParameterVariant> itp = parameters.iterator();
+    Iterator<rcl_interfaces.msg.ParameterDescriptor> itpd = descriptors.iterator();
+    synchronized (parametersMutex) {
```
Following my previous comment, this could become
```suggestion
    synchronized (this) {
```

jacobperron commented on rcljava/src/main/java/org/ros2/rcljava/node/NodeImpl.java:
```diff
@@ -392,103 +401,299 @@ public final String getName() {
+    // and undeclared simultaneously.
+    for (String s : parametersToDeclare) {
+      if (parametersToUndeclare.contains(s)) {
+        throw new IllegalArgumentException("Trying to both declare and undeclare the same parameter name");
```
Consider adding the name of the parameter to the exception message to help the user discover where the error is.

clalancette replied:
Good idea.  Done throughout the code in 36daa72840b68189f0c6d7d763440c2395ec0d46

jacobperron commented on rcljava/src/main/java/org/ros2/rcljava/node/NodeImpl.java:
```diff
@@ -392,103 +401,299 @@ public final String getName() {
+    // Undeclare any parameters that have their type set to NOT_SET
+    for (String s : parametersToUndeclare) {
+      if (!hasParameter(s)) {
+        throw new IllegalArgumentException("Parameter is not declared");
```
Same comment; consider adding the name of the parameter causing the issue.

clalancette replied:
36daa72840b68189f0c6d7d763440c2395ec0d46

jacobperron commented on rcljava/src/main/java/org/ros2/rcljava/node/NodeImpl.java:
```diff
@@ -392,103 +401,299 @@ public final String getName() {
+
+    // Check to make sure that a parameter isn't both trying to be declared
+    // and undeclared simultaneously.
+    for (String s : parametersToDeclare) {
```
nitpick: avoid single letter variables names.

`parameterName` is easier for me to parse than `s`. This comment applies throughout.

clalancette replied:
Fixes throughout the code in 36daa72840b68189f0c6d7d763440c2395ec0d46

jacobperron commented on rcljava/src/main/java/org/ros2/rcljava/node/NodeImpl.java:
```diff
@@ -392,103 +401,299 @@ public final String getName() {
+      return new rcl_interfaces.msg.ParameterDescriptor().setName(name);
+    }
+
+    throw new IllegalArgumentException("Parameter is not declared");
```
Add the name of the parameter to the exception message.

clalancette replied:
36daa72840b68189f0c6d7d763440c2395ec0d46

jacobperron commented on rcljava/src/main/java/org/ros2/rcljava/node/Node.java:
```diff
@@ -175,15 +177,39 @@
 
   String getName();
 
+  void declareParameter(ParameterVariant parameter);
```
rclcpp also returns the value of the declared parameter. If that's possible to do in rcljava, I think it would be convenient when writing application code.

clalancette replied:
I don't feel strongly about it either.  Given the API I've chosen here (taking in a ParameterVariant), the caller already had to do that, so it is somewhat redundant.  But for consistency with the rclcpp and rclpy APIs, I also now return it.  36daa72840b68189f0c6d7d763440c2395ec0d46

jacobperron commented on rcljava/src/test/java/org/ros2/rcljava/node/NodeParametersTest.java:
```diff
@@ -0,0 +1,383 @@
+    assertTrue(node.hasParameter("foo"));
+  }
+
+  @Test(expected = IllegalArgumentException.class)
```
We don't have to do it as part of this PR, but we should consider adding custom exceptions for parameter failure cases.
If you agree, I can make a follow-up ticket to track it.

clalancette replied:
Yeah, completely agreed.  Custom exceptions would make this much nicer for the user.  I also agree with making a follow-up ticket, since I don't want to make this already large PR even larger.  I'll go ahead and do that.

clalancette replied:
Oh, except we can't because this is a fork.  Not sure how to proceed here.

jacobperron replied:
I'll make a note locally for after this is upstreamed. We could also add a TODO in the code somewhere.

clalancette replied:
Good call about the TODO.  I've added one now.

jacobperron commented on rcljava/src/test/java/org/ros2/rcljava/node/NodeParametersTest.java:
```diff
@@ -0,0 +1,383 @@
+    assertTrue(node.hasParameter("foo"));
+    List<ParameterVariant> results = node.getParameters(new ArrayList<String>(Arrays.asList("foo")));
+    assertEquals(1, results.size());
+    assertEquals("foo", results.get(0).getName());
```
Can we test getting the value too?

clalancette replied:
Good idea.  Done in 36daa72840b68189f0c6d7d763440c2395ec0d46

jacobperron commented on rcljava/src/test/java/org/ros2/rcljava/node/NodeParametersTest.java:
```diff
@@ -0,0 +1,383 @@
+    node.declareParameter(new ParameterVariant("doh.baz", true));
+
+    Set<ParameterVariant> params = node.getParametersByPrefix("foo");
+    assertEquals(2, params.size());
```
Might be worth asserting that we got the right two.

clalancette replied:
Good call.  In fact, this lead me to completely reexamine getParameterByPrefix and realize that I had it completely wrong.  We really want to return a Map here to be compatible with the rclcpp and rclpy APIs.  Done in 36daa72840b68189f0c6d7d763440c2395ec0d46

jacobperron reviewed (commented):
Awesome stuff! Keeping logical changes per commit made it easy to review :+1: 

The only commit that seems completely unrelated is 5533b40, we could probably upstream that right away (e.g. to the `dashing` branch).

jacobperron commented on rcljava/src/main/java/org/ros2/rcljava/node/Node.java:
```diff
@@ -175,15 +177,361 @@
+   * Whatever name and value is in the given ParameterVariant will be used
+   * as the name and initial value of the parameter.
+   * If successful, the passed in ParameterVariant is returned.
+   * While the caller already has this information, this makes the API
```
nitpick: I don't think documenting "compatibility" with other client libraries is necessary. It's possible for there to be divergences between them, and then this statement might become false.

clalancette replied:
True.  I removed this now.

jacobperron commented on rcljava/src/main/java/org/ros2/rcljava/node/Node.java:
```diff
@@ -175,15 +177,361 @@
+   * set then an IllegalArgumentException is thrown.
+   *
+   * @param parameters The parameters to set.
+   * @param descriptors The descriptors to set.
```
nit: Document @return.

clalancette replied:
Done now.

jacobperron commented on rcljava/src/main/java/org/ros2/rcljava/node/Node.java:
```diff
@@ -175,15 +177,361 @@
+   *
+   * An empty string for the prefix will match all parameters.
+   *
+   * If no parameters with the prefix are found, the empty Set will be returned.
```
```suggestion
   * If no parameters with the prefix are found, the empty Map will be returned.
```

clalancette replied:
Oops, good call.

jacobperron reviewed (approved):
LGTM, just a few nitpicks

clalancette commented:
All right, I've now fixed up the remaining comments, and rebased the whole thing so it is a logical set of patches.  @jacobperron I don't have permissions on this repo to merge, so please go ahead when you are ready.  I'll suggest a "Rebase and Merge" rather than a "Squash and Merge" to keep the logical separation between patches.
